apps/communications/contacts/js/bootstrap.js

```diff
@@ -0,0 +1,423 @@
+    if (!container) {
+      console.warning('Could not apply cached content to ' +
+                      _cache.containerId);
+      return Promise.resolve();
+    }
+
+    return appendNodesToContainer(container, _cache.content).then(() => {
+      _cache.content = null;
+      _cache.active = true;
+      _caches.set(aCacheId, _cache);
+      return Promise.resolve();
+    });
+  };
+
+  function undoCache(aCacheId) {
+    if (!CACHE_ENABLED || !_caches.has(aCacheId)) {
+      return;
+    }
+
+    var _cache = _caches.get(aCacheId);
+
+    if (!_cache.containerId) {
+      return;
+    }
+
+    var container = document.getElementById(_cache.containerId);
+    if (!container) {
+      console.warning('Could not remove cached content from ' +
+                      _cache.containerId);
+      return;
+    }
+
+    removeCacheFromContainer(container);
+  }
+
+  var Cache = {
+    get active() {
+      return CACHE_ENABLED && _caches.get(FIRST_CHUNK).active;
+    },
+
+    get enabled () {
+      return CACHE_ENABLED;
+    },
+
+    get contacts() {
+      return (CACHE_ENABLED && _cachedContacts) ? _cachedContacts.keys()
+                                                : null;
+    },
+
+    get length() {
+      return (CACHE_ENABLED && _cachedContacts) ? _cachedContacts.size
+                                                : 0;
+    },
+
+    get headers() {
+      return (CACHE_ENABLED && _cachedHeaders) ? _cachedHeaders
+                                               : {};
+    },
+
+    // For testing purposes only.
+    get _rawContent() {
+      return _caches.get(FIRST_CHUNK).content;
+    },
+
+    set firstChunk(aContent) {
+      if (!CACHE_ENABLED || !aContent) {
+        return;
+      }
+      setCache({
+        id: FIRST_CHUNK,
+        content: aContent
+      });
+    },
+
+    set oneviction(aCallback) {
+      if (!CACHE_ENABLED) {
+        return;
+      }
+      addObserver('oneviction', aCallback);
+    },
+
+    hasContact: function(aUuid) {
+      if (!CACHE_ENABLED) {
+        return;
+      }
+      return (_cachedContacts && _cachedContacts.has(aUuid) ||
+              _cachedFavorites && _cachedFavorites.has(aUuid));
+    },
+
+    getContact: function(aUuid) {
+      if (!CACHE_ENABLED || !this.hasContact(aUuid)) {
+        return;
+      }
+      var contact;
+      // We should get each contact once while rendering the contacts list
+      // to see if what we have in the cache is different to what we have in
+      // the contacts source (most likely mozContacts). Removing the contact
+      // entry from the map allow us to easily check if we have any contact in
+      // the cache that was deleted from the original source and so it needs
+      // to be removed from the view.
+
+      if (_cachedContacts.has(aUuid)) {
+        contact = _cachedContacts.get(aUuid);
+        _cachedContacts.delete(aUuid);
+      } else if (_cachedFavorites.has(aUuid)) {
```

Comment:
Should be this `else if` an `else`?
Just thinking on a favorite contact that is on the first chunck, could happen in theory.